File: leetcode/py/p202-happy-number.py
# LeetCode 202. Happy Number.


def sum_square_digits(n) -> int:
    total = 0
    while n != 0:
        n, m = divmod(n, 10)
        total += m**2
    return total


class Solution:
    def isHappy(self, n: int) -> bool:
        intermediate_solutions = set()
        while n != 1:
            # print(f"{n}\n")
            n = sum_square_digits(n)
            if n in intermediate_solutions:
                return False
            intermediate_solutions.add(n)
        return True


def test_sum_square_digits():

    test_cases = (
        (1, 1),
        (2, 4),
        (9, 81),
        (10, 1),
        (19, 82),
        (91, 82),
    )
    for n, exp in test_cases:
        act = sum_square_digits(n)
        if exp != act:
            print(f"fail, sum_square_digits(n: {n}), exp: {exp}, act: {act}")


def test_isHappy():
    test_cases = (
        (1, True),
        (2, False),
        (19, True),
    )
    solution = Solution()
    for n, exp in test_cases:
        act = solution.isHappy(n)
        if exp != act:
            print(f"fail, isHappy(n: {n}), exp: {exp}, act: {act}")


if __name__ == "__main__":
    test_sum_square_digits()
    test_isHappy()

File: leetcode/cpp/p202-happy-number.cpp
// LeetCode 202. Happy Number.

#include <iostream>
#include <unordered_set>

static int sum_squares(int n)
{
    int total = 0;
    while (n != 0) {
        const auto nmod10 = n % 10;
        total += nmod10 * nmod10;
        n = n / 10;
    }
    return total;
}

class Solution {
public:
    bool isHappy(int n)
    {
        // Actually a vector might be faster.
        std::unordered_set<int> us;

        while (n != 1) {
            n = sum_squares(n);
            if (us.contains(n)) {
                return false;
            }
            us.insert(n);
        }
        return true;
    }
};

void test_sum_squares()
{
    struct TestCase {
        int n;
        int exp;
    };
    const TestCase test_cases[] = {
        { 1, 1 },
        { 2, 4 },
        { 9, 81 },
        { 19, 82 },
        { 91, 82 },
        { 91, 82 },
        { 2147483647, 260 },
    };
    for (const auto& tc : test_cases) {
        const auto act = sum_squares(tc.n);
        if (tc.exp != act) {
            std::cerr << "fail, n: " << tc.n
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}

void test_is_happy()
{
    struct TestCase {
        int n;
        bool exp;
    };
    const TestCase test_cases[] = {
        { 1, true },
        { 2, false },
        { 19, true },
        { 2147483647, false },
    };
    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.isHappy(tc.n);
        if (tc.exp != act) {
            std::cerr << "fail, n: " << tc.n
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}

int main()
{
    test_sum_squares();
    test_is_happy();
    return 0;
}
